ie-solver/helpers.cpp

```diff
@@ -67,35 +104,147 @@ double boundary_integral_solve(const ie_solver_config& config,
     skel_times->push_back(elapsed / TIMING_ITERATIONS);
     return -1;
   }
+  ie_solver_tools.skeletonize(kernel, &quadtree);
 
   ie_Mat f = config.boundary->boundary_values;
-  ie_Mat phi(config.num_boundary_points * solution_dimension, 1);
-  ie_solver_tools.skeletonize(kernel, &quadtree);
-  ie_solver_tools.solve(kernel, quadtree, &phi, f);
+  ie_Mat mu(config.num_boundary_points * solution_dimension, 1);
+
+
+
+  ie_Mat alpha(3, 1);
+  ie_Mat U(config.num_boundary_points * solution_dimension, 3);
+  ie_Mat Psi(3, config.num_boundary_points * solution_dimension);
+  ie_Mat ident(3, 3);
+  ident.set(0, 0, 1);
+  ident.set(1, 1, 1);
+  ident.set(2, 2, 1);
+  initialize_U_mat(config.boundary->points, &U);
+  initialize_Psi_mat(config.boundary.get(), &Psi);
+
+  // std::vector<unsigned int> alldofs;
+  // for (int i = 0; i < mu.height(); i++) {
+  //   alldofs.push_back(i);
+  // }
+  // ie_Mat whole = kernel(alldofs, alldofs);
+  // ie_Mat HUGE(alldofs.size() + 3, alldofs.size() + 3);
+  // HUGE.set_submatrix(alldofs, alldofs, whole);
+  // std::vector<unsigned int> others;
+  // others.push_back(alldofs.size());
+  // others.push_back(alldofs.size() + 1);
+  // others.push_back(alldofs.size() + 2);
+  // HUGE.set_submatrix(others, alldofs, Psi);
+  // HUGE.set_submatrix(alldofs, others, U);
+  // ident *= -1;
+  // HUGE.set_submatrix(others, others, ident);
+  // ident *= -1;
+  // std::cout << "cond num at start = " << whole.condition_number() << " " <<
+  //           HUGE.condition_number() << std::endl;
+
+  // ie_Mat big_vec(alldofs.size() + 3,1);
+  // std::vector<unsigned int> ZERO;
+  // ZERO.push_back(0);
+  // big_vec.set_submatrix(alldofs, ZERO, f);
+  // ie_Mat phi(alldofs.size(), 1);
+  // HUGE.left_multiply_inverse(big_vec, &phi);
+  // mu = phi(alldofs, ZERO);
+  // alpha = phi(others, ZERO);
+
+  ie_Mat U1, U2, U3;
+  std::vector<unsigned int> rows, cols;
+  for (unsigned int i = 0; i < U.height(); i++) {
+    rows.push_back(i);
+  }
+  cols.push_back(0);
+  U1 = U(rows, cols);
+  cols.clear();
+  cols.push_back(1);
+  U2 = U(rows, cols);
+  cols.clear();
+  cols.push_back(2);
+  U3 = U(rows, cols);
+
+  ie_Mat Dinv_U1(config.num_boundary_points * solution_dimension, 1);
+  ie_Mat Dinv_U2(config.num_boundary_points * solution_dimension, 1);
+  ie_Mat Dinv_U3(config.num_boundary_points * solution_dimension, 1);
+  ie_solver_tools.solve(kernel, quadtree, &Dinv_U1, U1);
+  ie_solver_tools.solve(kernel, quadtree, &Dinv_U2, U2);
+  ie_solver_tools.solve(kernel, quadtree, &Dinv_U3, U3);
+
+  ie_Mat Dinv_U(config.num_boundary_points * solution_dimension, 3);
+  cols.clear();
+  cols.push_back(0);
+  Dinv_U.set_submatrix(rows, cols, Dinv_U1);
+  cols.clear();
+  cols.push_back(1);
+  Dinv_U.set_submatrix(rows, cols, Dinv_U2);
+  cols.clear();
+  cols.push_back(2);
+  Dinv_U.set_submatrix(rows, cols, Dinv_U3);
+
+  ie_Mat Psi_Dinv_U(3, 3);
+  ie_Mat::gemm(NORMAL, NORMAL, 1., Psi, Dinv_U, 0., &Psi_Dinv_U);
+
+  ie_Mat S = ident;
```

Comment:
Yeah, implement the other matrix operators too so that these *=s aren't the only tool available